src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -401,22 +401,24 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
         /// Gets a new local symbol with the given TypeWithAnnotations as the new type. This
         /// type should be identical to the original except for nullability.
         /// </summary>
-        internal LocalSymbol WithAnalyzedType(TypeWithAnnotations analyzedType)
+        internal LocalSymbol WithContainingAndAnalyzedTypes(Symbol containingSymbol, TypeWithAnnotations analyzedType)
         {
+            Debug.Assert(Symbol.Equals(ContainingSymbol, containingSymbol, TypeCompareKind.AllNullableIgnoreOptions | TypeCompareKind.IgnoreTupleNames));
             Debug.Assert(analyzedType.Equals(TypeWithAnnotations, TypeCompareKind.AllNullableIgnoreOptions | TypeCompareKind.IgnoreTupleNames));
-            if (analyzedType.Equals(TypeWithAnnotations, TypeCompareKind.ConsiderEverything))
+            if (analyzedType.Equals(TypeWithAnnotations, TypeCompareKind.ConsiderEverything) &&
+                Symbol.Equals(ContainingSymbol, containingSymbol, TypeCompareKind.ConsiderEverything))
             {
                 return this;
             }
 
-            var cloned = CloneWithoutType();
+            var cloned = CloneWithoutType(containingSymbol);
             cloned.SetTypeWithAnnotations(analyzedType);
             return cloned;
         }
 
-        protected virtual SourceLocalSymbol CloneWithoutType()
+        protected virtual SourceLocalSymbol CloneWithoutType(Symbol containingSymbol)
         {
-            return new SourceLocalSymbol(_containingSymbol, _scopeBinder, allowRefKind: false, _typeSyntax, _identifierToken, _declarationKind);
+            return new SourceLocalSymbol(containingSymbol, _scopeBinder, allowRefKind: false, _typeSyntax, _identifierToken, _declarationKind);
```

Comment:
>return new SourceLocalSymbol(containingSymbol, _scopeBinder, allowRefKind: false, _typeSyntax, _identifierToken, _declarationKind); [](start = 12, length = 131)

I do not like the idea of cloning symbols that take binders, that means they potentially could binding something, but that should not happen. I think we should create thin wrappers that delegate to the original ones everything but container and type.